File: crates/axiotask-app/ui/src/__tests__/taskTree.test.js
import { describe, it, expect } from "vitest";
import { isSubtask, hasSubtasks, canAddSubtask, canNestUnder } from "../taskTree.js";

// Strict two-level tree (invariant #1): parents may have subtasks, subtasks may
// not. These predicates guard every mutation path against a third level.
const parent = { id: "p", parent_id: null };
const otherParent = { id: "p2", parent_id: null };
const sub = { id: "s", parent_id: "p" };
const childless = { id: "c", parent_id: null };
const tasks = [parent, otherParent, sub, childless];

describe("taskTree two-level invariant", () => {
  it("isSubtask: only tasks with a parent_id are subtasks", () => {
    expect(isSubtask(parent)).toBe(false);
    expect(isSubtask(sub)).toBe(true);
    expect(isSubtask(null)).toBe(false);
  });

  it("hasSubtasks: true only when some task points at the id", () => {
    expect(hasSubtasks("p", tasks)).toBe(true);
    expect(hasSubtasks("c", tasks)).toBe(false);
    expect(hasSubtasks("s", tasks)).toBe(false);
  });

  describe("canAddSubtask — a subtask can't gain a subtask", () => {
    it("allows adding under a top-level task", () => {
      expect(canAddSubtask(parent)).toBe(true);
      expect(canAddSubtask(childless)).toBe(true);
    });
    it("refuses adding under a subtask (would be a 3rd level)", () => {
      expect(canAddSubtask(sub)).toBe(false);
    });
    it("refuses when the parent is missing", () => {
      expect(canAddSubtask(null)).toBe(false);
      expect(canAddSubtask(undefined)).toBe(false);
    });
  });

  describe("canNestUnder — a task with subtasks can't become a subtask", () => {
    it("allows nesting a childless top-level task under another top-level task", () => {
      expect(canNestUnder("c", parent, tasks)).toBe(true);
    });
    it("refuses nesting a task that already has subtasks (its kids would be 3rd level)", () => {
      expect(canNestUnder("p", otherParent, tasks)).toBe(false);
    });
    it("refuses nesting under a subtask (would be a 3rd level)", () => {
      expect(canNestUnder("c", sub, tasks)).toBe(false);
    });
    it("refuses nesting under a missing parent", () => {
      expect(canNestUnder("c", null, tasks)).toBe(false);
    });
    it("refuses nesting a task under itself", () => {
      expect(canNestUnder("p", parent, tasks)).toBe(false);
    });
  });
});

File: test/model/task_tree_test.dart
// Unit layer — the enumerated `taskTree.test.js` cases ported 1:1.
//
// Protects invariant #1 (strict two levels). Every mutation path routes through
// these predicates; if canNestUnder/canAddSubtask drifted, a subtask could gain
// a subtask (a third level) — the exact corruption the invariant forbids.

import 'package:axiotask/src/model/task.dart';
import 'package:axiotask/src/model/task_tree.dart';
import 'package:flutter_test/flutter_test.dart';

Task t(String id, {String? parent}) => Task(
  id: id,
  parent: parent,
  position: '1',
  title: id,
  status: TaskStatus.needsAction,
  updated: 'u',
);

void main() {
  final parent = t('p');
  final otherParent = t('p2');
  final sub = t('s', parent: 'p');
  final childless = t('c');
  final tasks = [parent, otherParent, sub, childless];

  group('isSubtask', () {
    test('only tasks with a parent are subtasks', () {
      expect(isSubtask(parent), isFalse);
      expect(isSubtask(sub), isTrue);
      expect(isSubtask(null), isFalse);
    });
  });

  group('hasSubtasks', () {
    test('true only when some task points at the id', () {
      expect(hasSubtasks('p', tasks), isTrue);
      expect(hasSubtasks('c', tasks), isFalse);
      expect(hasSubtasks('s', tasks), isFalse);
    });
  });

  group('canAddSubtask — a subtask cannot gain a subtask', () {
    test('allows adding under a top-level task', () {
      expect(canAddSubtask(parent), isTrue);
      expect(canAddSubtask(childless), isTrue);
    });
    test('refuses adding under a subtask (would be a 3rd level)', () {
      expect(canAddSubtask(sub), isFalse);
    });
    test('refuses when the parent is missing', () {
      expect(canAddSubtask(null), isFalse);
    });
  });

  group('canNestUnder — a task with subtasks cannot become a subtask', () {
    test(
      'allows nesting a childless top-level task under another top-level',
      () {
        expect(canNestUnder('c', parent, tasks), isTrue);
      },
    );
    test('refuses nesting a task that already has subtasks', () {
      expect(canNestUnder('p', otherParent, tasks), isFalse);
    });
    test('refuses nesting under a subtask (would be a 3rd level)', () {
      expect(canNestUnder('c', sub, tasks), isFalse);
    });
    test('refuses nesting under a missing parent', () {
      expect(canNestUnder('c', null, tasks), isFalse);
    });
    test('refuses nesting a task under itself', () {
      expect(canNestUnder('p', parent, tasks), isFalse);
    });
  });
}
